### ISSUE TITLE
`otherdims` is type unstable

### ISSUE THREAD

**Issue by Datseris** [rafaqz/DimensionalData.jl#164]
```

julia> A = DimArray(rand(40, 50), (X, Y));

julia> @inferred otherdims(A, (X,))
ERROR: return type Tuple{Y{Base.OneTo{Int64},NoIndex,Nothing}} does not match inferred return type Tuple
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] top-level scope at none:1
```

on 0.12. I've also tested with much more complicated `A` and dimension types.

**Comment by rafaqz** [comment]
That's because you are using the `(X, Y)`. Use `(X(), Y())` if you want performance.

We could look at automatically constructing them before they get to deep, which might fix the problem.

**Comment by rafaqz** [comment]
It does :) Nice when it only takes two characters.

**Comment by Datseris** [comment]
great!

It would be great to fix at source (if possible) to not necessitate using `()`, not because I can't do it, but because the result of this function is called by other subesequent functions by the user within our climatetools framework, and it is not as easy for a simple user to understand this small, but critical difference.

**Comment by Datseris** [comment]
(mostly because other things like `mean(A; dims = Time)` are called without `()` with no problem)

**Comment by rafaqz** [comment]
So actually it works for a single dimension but not types in a tuple. Because:

```julia
julia> typeof((X, Y))
Tuple{UnionAll,UnionAll}
```

It's a good reason to allow `dims(A, X, Y)` syntax, as it won't have this problem.

**Comment by rafaqz** [comment]
It's just base julia behaviour. I knew this but got hopeful for a second there...
```julia
julia>     @inferred map(x -> x(), (X, Y))
ERROR: return type Tuple{X{Colon,AutoMode{AutoOrder},Nothing},Y{Colon,AutoMode{AutoOrder},Nothing}} does not match inferred return type Tuple{Any,Any}
```

**Comment by Datseris** [comment]
Kind of related: should the docstrings in https://github.com/rafaqz/DimensionalData.jl/pull/161/files  be updated to use `X()` instead of `X`, etc.?

**Comment by rafaqz** [comment]
It's hard to say, for the average use case the type stability is irrelevant. It only matters in a hot loop.

We should probably resolve it by letting you write `dims(A, X, Y)`, `mode(A, X, Ti)` etc with arguments. It's type stable that way, and maybe the nicest syntax.

**Comment by Datseris** [comment]
I agree.

**Comment by rafaqz** [comment]
Closed, will be resolved with #180
